src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Instance.hs

```diff
@@ -28,75 +38,85 @@ import           ZkFold.Symbolic.Data.Eq
+    r1 + r2 = circuitN $ do
+        is <- runCircuit r1
+        js <- runCircuit r2
+        for (Z.zip is js) $ \(i, j) -> newAssigned (\x -> x i + x j)
+
+instance (Arithmetic a, Scale c a, KnownNat n) => Scale c (ArithmeticCircuit n a) where
+    scale c r = circuitN $ do
+        is <- runCircuit r
+        for is $ \i -> newAssigned (\x -> (c `scale` one :: a) `scale` x i)
+
+instance (Arithmetic a, KnownNat n) => AdditiveMonoid (ArithmeticCircuit n a) where
+    zero = circuitN $ Vector <$> replicateM (Haskell.fromIntegral $ value @n) (newAssigned (const zero))
+
+instance (Arithmetic a, KnownNat n) => AdditiveGroup (ArithmeticCircuit n a) where
+    negate r = circuitN $ do
+        is <- runCircuit r
+        for is $ \i -> newAssigned (\x -> negate (x i))
+
+    r1 - r2 = circuitN $ do
+        is <- runCircuit r1
+        js <- runCircuit r2
+        for (Z.zip is js) $ \(i, j) -> newAssigned (\x -> x i - x j)
+
+instance (Arithmetic a, KnownNat n) => MultiplicativeSemigroup (ArithmeticCircuit n a) where
+    r1 * r2 = circuitN $ do
+        is <- runCircuit r1
+        js <- runCircuit r2
+        for (Z.zip is js) $ \(i, j) -> newAssigned (\x -> x i * x j)
+
+instance (Arithmetic a, KnownNat n) => Exponent (ArithmeticCircuit n a) Natural where
     (^) = natPow
 
-instance Arithmetic a => MultiplicativeMonoid (ArithmeticCircuit a) where
+instance (Arithmetic a, KnownNat n) => MultiplicativeMonoid (ArithmeticCircuit n a) where
     one = mempty
 
-instance (Arithmetic a, FromConstant b a) => FromConstant b (ArithmeticCircuit a) where
-    fromConstant c = embed (fromConstant c)
+-- TODO: The constant will be replicated in all outputs. Is this the desired behaviour?
+instance (Arithmetic a, FromConstant b a, KnownNat n) => FromConstant b (ArithmeticCircuit n a) where
+    fromConstant c = embedAll (fromConstant c)
 
-instance Arithmetic a => Semiring (ArithmeticCircuit a)
+instance (Arithmetic a, KnownNat n) => Semiring (ArithmeticCircuit n a)
 
-instance Arithmetic a => Ring (ArithmeticCircuit a)
+instance (Arithmetic a, KnownNat n) => Ring (ArithmeticCircuit n a)
 
-instance Arithmetic a => Exponent (ArithmeticCircuit a) Integer where
+instance (Arithmetic a, KnownNat n) => Exponent (ArithmeticCircuit n a) Integer where
     (^) = intPowF
 
-instance Arithmetic a => Field (ArithmeticCircuit a) where
+instance (Arithmetic a, KnownNat n) => Field (ArithmeticCircuit n a) where
     finv = invertC
-    rootOfUnity n = embed <$> rootOfUnity n
-
-instance Arithmetic a => BinaryExpansion (ArithmeticCircuit a) where
-    binaryExpansion r = circuits $ runCircuit r >>= expansion (numberOfBits @a)
-    fromBinary bits = circuit $ Haskell.traverse runCircuit bits >>= horner
-
-instance Arithmetic a => DiscreteField' (ArithmeticCircuit a) where
+    rootOfUnity n = embedAll <$> rootOfUnity n
+
+-- TODO: The only implementation that seems to make sense is when there is only one output.
+-- It is true?
```

Comment:
Yes, probably. It would be too unintuitive to consider expansions of multiple numbers at the same time.

Reply:
Well, given the laws for `BinaryExpansion`, an `instance Semiring (ArithmeticCircuit n a)` hints that there is one instance where in `binaryExpansion r :: [ArithmeticCircuit n a]` the `i`-th output of a `j`-th circuit is a `j`-th bit of an `i`-th output of `r`.